labrad/crypto.py

```diff
@@ -0,0 +1,134 @@
+    if cert_string is None:
+        cert_string = load_cert(hostname)
+    if cert_string is None:
+        logging.info('No certificate found for host "{}" in {}. Will use '
+                     'system certs to verify tls'.format(hostname, CERTS_PATH))
+        return ssl.optionsForClientTLS(hostname.decode('utf-8'))
+    else:
+        cert = ssl.Certificate.loadPEM(cert_string)
+        return ssl.optionsForClientTLS(hostname.decode('utf-8'), cert)
+
+
+def save_cert(hostname, cert_string):
+    """Save the given self-signed certificate as a trusted certificate.
+
+    Args:
+        hostname (string): The hostname for this certificate. Must match the
+            canonical name attribute (CN) in the certificate itself. We do not
+            check this here, but if it is not the case, TLS negotiation using
+            this cert will fail.
+        cert_string (string): Certificate in PEM format.
+    """
+    # sanity check that this is a valid PEM string
+    cert = ssl.Certificate.loadPEM(cert_string)
+
+    fname = hostname + '.cert'
+    full_name = os.path.join(CERTS_PATH, fname)
+    if not os.path.exists(CERTS_PATH):
+        os.makedirs(CERTS_PATH)
+    with open(full_name, 'w') as f:
+        f.write(cert_string)
+
+
+def load_cert(hostname):
+    """Load a trusted certificate for the given hostname.
+
+    Args:
+        hostname(string): The hostname for which we want a certificate.
+
+    Returns (string | None):
+        A certificate for the specified hostname as a PEM-encoded string. If no
+        certificate is found for the given hostname, we return None.
+    """
+    if not os.path.exists(CERTS_PATH):
+        return None
+    for fname in os.listdir(CERTS_PATH):
+        full_name = os.path.join(CERTS_PATH, fname)
+        if not os.path.isfile(full_name):
+            continue
+        host, ext = os.path.splitext(fname)
+        if host == hostname and ext == '.cert':
+            with open(full_name) as f:
+                return f.read()
+    else:
+        return None
+
+
+def fingerprint(cert_string, digest=hashlib.sha1):
+    """Get a human-readable digest of the given certificate string.
+
+    Certificate fingerprints are computed from the DER-encoded (binary) form of
+    the certificate, while PEM is a text format containing the base64-encoded
+    DER bytes. So we strip off boilerplate/whitespace and base64-decode before
+    computing the digest.
+
+    Args:
+        cert_string (string): PEM-encoded certificate.
+        digest: the hashlib algorithm to use; a callable that returns an
+            instance of a message digest algorithm.
+
+    Returns (string):
+        Human-readable hex version of the certificate digest in the form:
+        'AA:BB:CC:...:FF'.
+    """
+    cert_b64 = (cert_string.replace('-----BEGIN CERTIFICATE-----', '')
+                           .replace('-----END CERTIFICATE-----', '')
+                           .replace('\n', ''))
+    cert_bytes = base64.b64decode(cert_b64)
```

Comment:
You should strip stuff before and after the BEGIN CERTIFICATE and END CERTIFICATE demarcations.